### PR TITLE
KAFKA-14073; Log the reason for snapshot

### PR DESCRIPTION
**Pull request by ashmeet13**
When a snapshot is taken it is due to either of the following reasons - 
1. Max bytes were applied
2. Metadata version was changed

Once the snapshot process is started, it will log the reason that initiated the process.

Updated existing tests to include code changes required to log the reason. I was not able to check the logs when running tests - could someone guide me on how to enable logs when running a specific test case. 

Example logs after the changes - 
```
[2022-07-25 14:34:39,769] INFO [Controller 3000] Generating a snapshot that includes (epoch=1, offset=0) after 91 committed bytes since the last snapshot, because max bytes exceeded. (org.apache.kafka.controller.QuorumController:1328)
```

```
[2022-07-25 14:40:51,783] INFO [BrokerMetadataSnapshotter id=2] Creating a new snapshot at offset 5 because metadata version changed... (kafka.server.metadata.BrokerMetadataSnapshotter:66)
```
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ashmeet13** [comment]
@dengziming since you raised the JIRA ticket for [KAFKA-14073](https://issues.apache.org/jira/browse/KAFKA-14073), would please review this PR.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 1]
```diff
@@ -89,6 +89,12 @@ class BrokerMetadataListener(
    */
   private var _bytesSinceLastSnapshot: Long = 0L
 
+  /**
+   * The reason as to why we are calling maybeStartSnapshot, can be either
+   * MaxBytesExceeded or MetadataVersionChanged
+   */
+  private var _reasonForSnapshot: String = ""
```
Comment (dengziming) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: I think it's unnecessary to make "reason" a class variable, a local variable is necessary.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 2]
```diff
@@ -128,7 +134,15 @@ class BrokerMetadataListener(
   }
 
   private def shouldSnapshot(): Boolean = {
-    (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots) || metadataVersionChanged()
+    if (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots) {
+      _reasonForSnapshot = "MaxBytesExceeded"
```
Comment (dengziming) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: since the log is for users, we can use "max bytes exceeded" style format instead of camel case

**Review by dengziming — COMMENTED** [review]
Thank you @ashmeet13 , I left 2 comments, and there is another place we should also log the reason: `QuorumController.SnapshotGeneratorManager.createSnapshotGenerator`, you can find that it is also calling `raftClient.createSnapshot`.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala (hunk 1 above)
Comment (ashmeet13) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Thanks, I have made the change. I have handled it by returning a `Tuple` from the `shouldSnapshot` function.
I am new to Scala, if there is another recommended way of doing it please let me know!

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala (hunk 2 above)
Comment (ashmeet13) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Got it, made this change

**Comment by ashmeet13** [comment]
@dengziming I have made the changes requested. I have also made the changes in `QuorumController.maybeGenerateSnapshot` to log the reason for the snapshot being generated. 

`maybeGenerateSnapshot` seemed to be a better place than `QuorumController.SnapshotGeneratorManager.createSnapshotGenerator` since it already had logs for starting a snapshot.

Also, it seems like `QuorumController` only creates a snapshot when max bytes are exceeded, I couldn't find it calling the snapshot generator for the reason metadata version changed.

**Comment by ashmeet13** [comment]
Hi @dengziming, could you review this once more? I have made the requested changes.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 3]
```diff
@@ -119,16 +119,24 @@ class BrokerMetadataListener(
...
+        maybeStartSnapshot(snapshotReason)
       }
 
       _publisher.foreach(publish)
     }
   }
 
-  private def shouldSnapshot(): Boolean = {
-    (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots) || metadataVersionChanged()
+  private def shouldSnapshot(): (Boolean, String) = {
```
Comment (dengziming) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Here we can use an Option to represent the reason, similar to `QuorumFeatures.reasonNotSupported`.

**Review by dengziming — COMMENTED** [review]
Thanks for the update, left one comment. you can also run some integration test to see the log and paste them in the description.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala (hunk 3 above)
Comment (ashmeet13) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Thanks, I have made this change.

**Comment by ashmeet13** [comment]
@dengziming I have made the requested changes and added the sample logs in the description of the PR.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 4]
```diff
@@ -119,16 +119,24 @@ class BrokerMetadataListener(
...
       }
 
       _publisher.foreach(publish)
     }
   }
 
-  private def shouldSnapshot(): Boolean = {
-    (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots) || metadataVersionChanged()
+  private def shouldSnapshot(): Option[String] = {
+    if (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots) {
```
Comment (dengziming) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: I think it's possible that both these 2 conditions are satisfied at the same time, right? so maybe we should have another condition: "max bytes exceeded and metadata version changed", WDYT?

Comment (ashmeet13) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Makes sense, I wasn't sure whether this scenario was possible or not. Haven't dug deep yet. Should've clarified this before, my bad.

Have made the changes to accommodate this.

**Review by dengziming — APPROVED** [review]
LGTM,  ping @jsancio to have a look.

File: metadata/src/main/java/org/apache/kafka/controller/QuorumController.java [hunk 5]
```diff
@@ -1325,7 +1325,7 @@ private void maybeGenerateSnapshot(long batchSizeInBytes) {
                 snapshotRegistry.getOrCreateSnapshot(lastCommittedOffset);
             }
 
-            log.info("Generating a snapshot that includes (epoch={}, offset={}) after {} committed bytes since the last snapshot.",
+            log.info("Generating a snapshot that includes (epoch={}, offset={}) after {} committed bytes since the last snapshot, because max bytes exceeded.",
```
Comment (jsancio) [inline metadata/src/main/java/org/apache/kafka/controller/QuorumController.java]: Hmm. So it looks like in the Controller we don't generate a snapshot when the metadata version changes.

@mumrah How should we resolve this?

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 6]
```diff
@@ -119,16 +119,29 @@ class BrokerMetadataListener(
...
+        maybeStartSnapshot(shouldTakeSnapshot.get)
       }
 
       _publisher.foreach(publish)
     }
   }
 
-  private def shouldSnapshot(): Boolean = {
-    (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots) || metadataVersionChanged()
+  private def shouldSnapshot(): Option[String] = {
```
Comment (jsancio) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Let's have a stricter type. I suspect that we are going to add more reasons for generating snapshots. What do you think if we create a Java enum `SnapshotReason` in the `raft` module and `org.apache.kafka.snapshot` package? We can also changed the signature of `RaftClient#creaeSnapshot` to

```
Optional<SnapshotWriter<T>> createSnapshot(
    long committedOffset,
    int committedEpoch,
    long lastContainedLogTime,
    SnapshotReason reason
);
```

You can take a look at `LogStartOffsetIncrementReason` in `core/src/main/scala/kafka/log/UnifiedLog.scala` for an example in Scala. We want this as an enum in Java since it will be used in `KafkaRaftClient`.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataSnapshotter.scala [hunk 7]
```diff
@@ -63,7 +63,7 @@ class BrokerMetadataSnapshotter(
       )
       if (writer.nonEmpty) {
         _currentSnapshotOffset = image.highestOffsetAndEpoch().offset
-        info(s"Creating a new snapshot at offset ${_currentSnapshotOffset}...")
+        info(s"Creating a new snapshot at offset ${_currentSnapshotOffset} because ${snapshotReason}...")
```
Comment (jsancio) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataSnapshotter.scala]: What do you think about moving this log line to the `KafkaMetadataLog` type? That code is shared between the Controller and the Broker.

File: metadata/src/main/java/org/apache/kafka/controller/QuorumController.java (hunk 5 above)
Comment (jsancio) [reply metadata/src/main/java/org/apache/kafka/controller/QuorumController.java]: What do you think about moving this log line to the `KafkaMetadataLog` type? That code is shared between the Controller and the Broker.

**Review by jsancio — COMMENTED** [review]
Thanks for the improvements @ashmeet13

**Comment by ashmeet13** [comment]
Got it, working on these changes. Taking a bit of time to go through the code.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 8]
```diff
@@ -129,16 +129,17 @@ class BrokerMetadataListener(
     }
   }
 
-  private def shouldSnapshot(): Option[String] = {
+  private def shouldSnapshot(): Option[SnapshotReason] = {
     val metadataVersionHasChanged: Boolean = metadataVersionChanged()
     val maxBytesHaveExceeded: Boolean = (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots)
 
+    //TODO: Figure out how to handle multiple reasons for snapshot
```
Comment (jsancio) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: One solution is to return a `Set[SnapshotReason]` and pass the set when creating the snapshot.

**Comment by ashmeet13** [comment]
Hi @jsancio, I have defined an enum in the raft module for `SnapshotReason` and have used it in place of string messages that were being passed previously.

I am working on making the changes for `RaftClient` and `KafkaMetadataLog`. These changes will log the snapshot reason in `createNewSnapshot` function of `KafkaMetadataLog`.

I had two implementation queries, could you please help with them - 
1. How do we handle multiple reasons for starting a snapshot in an enum? 
2. I would also need to change the function signature of `createNewSnapshot()` in `KafkaMetadataLog` to accommodate the new `SnapshotReason` parameter. Below is the new signature. Would this be okay?

```
Optional<RawSnapshotWriter> createNewSnapshot(OffsetAndEpoch snapshotId, SnapshotReason reason);
```

**Comment by jsancio** [comment]
Thanks for the changes @ashmeet13 they look good in general.

>     1. How do we handle multiple reasons for starting a snapshot in an enum?

Instead of using `Optional[SnapshotReason]` we could use `Set[SnapshotReason]`. Depending on the layer we can associate different means to this set. When returned from `def shouldSnapshot()` the caller can assume that if the set is empty it means that it should not generate a snapshot.

>     2. I would also need to change the function signature of `createNewSnapshot()` in `KafkaMetadataLog` to accommodate the new `SnapshotReason` parameter. Below is the new signature. Would this be okay?
> ```java
> Optional<RawSnapshotWriter> createNewSnapshot(OffsetAndEpoch snapshotId, SnapshotReason reason);
> ```
How about:
 ```java
 Optional<RawSnapshotWriter> createNewSnapshot(OffsetAndEpoch snapshotId, Set<SnapshotReason> reasons);
 ```
If the set `reasons` is empty then the implementation of `createNewSnapshot` can assume that the reason is unknown.

What do you think?

**Comment by ashmeet13** [comment]
Thanks @jsancio, this is helpful. I'll go ahead and make these changes!

**Comment by jsancio** [comment]
> Thanks @jsancio, this is helpful. I'll go ahead and make these changes!

@ashmeet13 ,

I thought about this some more and I don't think we should include the reason for the snapshot in the `raft` module and types. In other words, I think we should:
1. Move the `SnapshotReason` type from `raft/src/main/java/org/apache/kafka/raft` to `metadata/src/main/java/org/apache/kafka/metadata/util`.
2. Keep the old signature for the `createNewSnapshot`. I suspect that we can undo all of the changes you made to the `raft` module.

I think this now because the reasons for creating a snapshot are specific to `metadata` and not `raft`. We have an interest in using the `raft` module and the KRaft protocol to solve other problems. I don't think we will have the same reasons for wanting to generate snapshot in those use cases.

We should still use the `SnapshotReason` type and I am just concern on where it should live. Thanks and excuse my earlier misguided advice.

**Comment by ashmeet13** [comment]
Hi @jsancio, just one another doubt - Does this mean we don't want to log the reason within `KafkaMetadataLog` and keep the logging as it is right now within `BrokerMetadataSnapshotter` and `QuorumController`?

**Comment by jsancio** [comment]
> Hi @jsancio, just one another doubt - Does this mean we don't want to log the reason within `KafkaMetadataLog` and keep the logging as it is right now within `BrokerMetadataSnapshotter` and `QuorumController`?

Yes. That's correct. Let's keep the logging we have now and extend the message to include the reason(s).

**Comment by ashmeet13** [comment]
Hi @jsancio, I have reverted the changes made to `RaftClient` and kept the logging withing `BrokerMetadataSnapshotter` and `QuorumController`

I have two questions, would really be helpful if you could guide me on this - 
1. We also take a snapshot in `org.apache.kafka.raft` within the class `ReplicatedCounter` under the `handleCommit` function.
I am facing an issue in being able to import the `SnapshotReason` from the `metadata/src/main/java/org/apache/kafka/metadata/util` package. Could you guide me on how I could fix this?

2. We take snapshots at two different places within `QuorumController` - One I was able to figure out the reason being `MaxBytesExceeded`. There is another function `beginWritingSnapshot` which also initiates a snapshot - I was not able to figure out who calls this function and what would the reason be in this scenario?

File: raft/src/main/java/org/apache/kafka/raft/ReplicatedCounter.java [hunk 9]
```diff
@@ -107,6 +107,8 @@ public synchronized void handleCommit(BatchReader<Integer> reader) {
             }
             log.debug("Counter incremented from {} to {}", initialCommitted, committed);
 
+            // A snapshot is being taken here too, not being able to -
+            // `import org.apache.kafka.metadata.utils.SnapshotReason`, figure out why?
```
Comment (dengziming) [inline raft/src/main/java/org/apache/kafka/raft/ReplicatedCounter.java]: The `raft` module only relies on the `clients` module so it can't import a class from the `metadata` module.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataSnapshotter.scala [hunk 10]
```diff
@@ -59,11 +60,19 @@ class BrokerMetadataSnapshotter(
...
+
+        var stringReasons: Set[String] = Set()
+
+        snapshotReasons.foreach(r => stringReasons += r.toString)
+
+        if (stringReasons.isEmpty){
+          stringReasons += SnapshotReason.UnknownReason.toString
+        }
+
+        info(s"Creating a new snapshot at offset ${_currentSnapshotOffset} because, ${stringReasons.mkString(" and ")}")
```
Comment (dengziming) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataSnapshotter.scala]: We directly use snapshotReasons.mkString(" and ") here and remove `stringReasons` and related code.

Comment (ashmeet13) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataSnapshotter.scala]: Made this change

File: raft/src/main/java/org/apache/kafka/raft/ReplicatedCounter.java (hunk 9 above)
Comment (ashmeet13) [reply raft/src/main/java/org/apache/kafka/raft/ReplicatedCounter.java]: I see, so will we ignore this snapshot request and not log the reason for it?

**Comment by ashmeet13** [comment]
Hi @dengziming @jsancio, can you help with a review on this PR. If it looks okay to merge I can clean up the PR and remove temporary comments added by me.

File: raft/src/main/java/org/apache/kafka/raft/ReplicatedCounter.java (hunk 9 above)
Comment (dengziming) [reply raft/src/main/java/org/apache/kafka/raft/ReplicatedCounter.java]: Here we already logged the reason: "the previoud snapshot includes {}", please remove the new comments and help to fix the typo "previoud".

**Review by dengziming — COMMENTED** [review]
LGTM, left one comments.

**Comment by ashmeet13** [comment]
Thank you @dengziming for the review. Made the changes to remove the comment I had added and fixed the typo in the log message.

File: metadata/src/main/java/org/apache/kafka/controller/QuorumController.java [hunk 11]
```diff
@@ -1914,6 +1915,7 @@ public CompletableFuture<List<CreatePartitionsTopicResult>> createPartitions(
         });
     }
 
+    // TODO: Figure out the reason as to why is a snapshot starting here? Who is calling beginWritingSnapshot()?
```
Comment (dengziming) [inline metadata/src/main/java/org/apache/kafka/controller/QuorumController.java]: This method is used in tests, please also remove this comment.

Comment (ashmeet13) [reply metadata/src/main/java/org/apache/kafka/controller/QuorumController.java]: Hi @dengziming, forgot to remove this comment. I have removed it now.
Thanks

**Review by dengziming — APPROVED** [review]
LGTM

**Comment by ashmeet13** [comment]
Thanks @dengziming for the review.
@jsancio could you please have a look at this PR?

**Comment by ashmeet13** [comment]
Hi @jsancio, bumping this PR up in case it got missed. Could please have a look at this?

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 12]
```diff
@@ -119,16 +120,29 @@ class BrokerMetadataListener(
...
 
       _publisher.foreach(publish)
     }
   }
 
-  private def shouldSnapshot(): Boolean = {
-    (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots) || metadataVersionChanged()
+  private def shouldSnapshot(): Set[SnapshotReason] = {
+    val metadataVersionHasChanged: Boolean = metadataVersionChanged()
+    val maxBytesHaveExceeded: Boolean = (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots)
```
Comment (jsancio) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Minor but we tend not to declare the type for private variables with simple expressions.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataSnapshotter.scala [hunk 13]
```diff
@@ -59,11 +60,11 @@ class BrokerMetadataSnapshotter(
       val writer = writerBuilder.build(
         image.highestOffsetAndEpoch().offset,
         image.highestOffsetAndEpoch().epoch,
-        lastContainedLogTime
-      )
+        lastContainedLogTime)
```
Comment (jsancio) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataSnapshotter.scala]: Minor but you can revert this change. It is okay to have the closing delimiter in a new line.

File: core/src/main/scala/kafka/server/metadata/MetadataSnapshotter.scala [hunk 14]
```diff
@@ -31,5 +32,5 @@ trait MetadataSnapshotter {
    *
    * @return                      True if we will write out a new snapshot; false otherwise.
    */
-  def maybeStartSnapshot(lastContainedLogTime: Long, image: MetadataImage): Boolean
+  def maybeStartSnapshot(lastContainedLogTime: Long, image: MetadataImage, reason: Set[SnapshotReason]): Boolean
```
Comment (jsancio) [inline core/src/main/scala/kafka/server/metadata/MetadataSnapshotter.scala:36]: Let's add the new parameter to the documentation above.

**Review by jsancio — COMMENTED** [review]
Looks good @ashmeet13 just some minor suggestions. Thanks for the fix and excuse my delayed response.

**Comment by jsancio** [comment]
> Hi @jsancio, I have reverted the changes made to `RaftClient` and kept the logging withing `BrokerMetadataSnapshotter` and `QuorumController`
> 
> I have two questions, would really be helpful if you could guide me on this -
> 
>     1. We also take a snapshot in `org.apache.kafka.raft` within the class `ReplicatedCounter` under the `handleCommit` function.
>        I am facing an issue in being able to import the `SnapshotReason` from the `metadata/src/main/java/org/apache/kafka/metadata/util` package. Could you guide me on how I could fix this?

`ReplicatedCounter` is mainly used to test the KRaft implementation. We should probably consider moving this type to `raft/src/test`.
> 
>     2. We take snapshots at two different places within `QuorumController` - One I was able to figure out the reason being `MaxBytesExceeded`. There is another function `beginWritingSnapshot` which also initiates a snapshot - I was not able to figure out who calls this function and what would the reason be in this scenario?

It looks like this method is only used by tests. I think it is okay for this to have an "unknown reason" for now.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala (hunk 12 above)
Comment (ashmeet13) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Removed the type declaration

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataSnapshotter.scala (hunk 13 above)
Comment (ashmeet13) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataSnapshotter.scala]: My bad, must have done this by mistake. Fixed this.

File: core/src/main/scala/kafka/server/metadata/MetadataSnapshotter.scala (hunk 14 above)
Comment (ashmeet13) [reply core/src/main/scala/kafka/server/metadata/MetadataSnapshotter.scala:36]: Added the `reason` as a `param` in the documentation.

**Comment by ashmeet13** [comment]
Thank you @jsancio for the review. Have made the changes.

For the following comment -
>It looks like this method is only used by tests. I think it is okay for this to have an "unknown reason" for now.

Currently the function in discussion has no logging taking place hence I haven't added the `UnknownReason` in. Should I update the function to log the same?

Edit - I have gone ahead and added the log the same way we are logging in the actual function within `QuorumController`. I can remove it if necessary.

**Comment by ashmeet13** [comment]
Hi @jsancio could you please review this PR?  Sorry for the multiple changes this has required.

**Comment by jsancio** [comment]
> Hi @jsancio could you please review this PR? Sorry for the multiple changes this has required.

Thanks for the changes and excuse the delays. LGTM in general. Restarted the build as it seems the JVM 17 and Scala 2.13 configuration failed to build.

**Comment by ashmeet13** [comment]
Thank you @jsancio, I see that the build for JVM 17 and JDK 2.13 has timed out again. What can be done from my end to fix this?

**Review by jsancio — APPROVED** [review]
LGTM. Thanks for the improvements @ashmeet13
